index.js

```diff
@@ -0,0 +1,563 @@
+    await paymentService.db.save({payment});
+    throw new BedrockError(
+      'Expected PayPal Status COMPLETED.',
+      Errors.Data, {payPalStatus: order.status}
+    );
+  }
+  const {purchase_units: paypalPurchases} = order;
+  const verifiedPurchase = verifyPurchase({paypalPurchases});
+  // Guards against the user re-using a previous order from PayPal.
+  const existingPayments = await paymentService.db.findAll(
+    {query: {paymentService: 'paypal', paymentServiceId: order.id}});
+  // there should be at most 1 existingPayments.
+  if(existingPayments.length > 1) {
+    // This could be problematic as the user whose card was
+    // charged might not get their product.
+    // This does prevent duplicate charges using the same PayPal order id.
+    throw new BedrockError(
+      'More than one Payment found for PayPal order.',
+      Errors.Duplicate, {payPalId: order.id}
+    );
+  }
+  const {total} = getTotalCost({order});
+  verifiedPurchase.totalCost = total;
+  return verifiedPurchase;
+};
+
+/**
+ * Creates a PayPal order server side with no payment
+ * information filled out.
+ *
+ * @param {object} options - Options to use.
+ * @param {Array<object>} options.purchase_units - A purchase_unit.
+ * @param {string} [options.intent = 'CAPTURE'] - The order intent.
+ *
+ * @returns {Promise<object>} The order object.
+ */
+const createGatewayPayment = async ({payment, intent = 'CAPTURE'}) => {
+  try {
+    const {id, amount: value, currency = 'USD'} = payment;
+    const amount = formatAmount({amount: {currency_code: currency, value}});
+    const purchase_units = [{
+      reference_id: id,
+      amount
+    }];
+    const {brandName, shippingPreference} = config.paypal;
+    const application_context = {
+      brand_name: brandName || 'bedrock-order',
+      shipping_preference: shippingPreference || 'NO_SHIPPING'
+    };
+    logger.debug('ORDER APPLICATION CONTEXT', {application_context});
+    const {headers, api} = await getOptions();
+    const url = `${api}/v2/checkout/orders`;
+    headers['Content-Type'] = 'application/json';
+    const options = {headers};
+    const body = {
+      intent,
+      purchase_units,
+      application_context
+    };
+    const {data} = await axios.post(url, body, options);
+    payment.paymentServiceId = data.id;
+    return {order: data, payment};
+  } catch(error) {
+    const {errorType, details} = formatAxiosError({error});
+    throw new BedrockError(
+      'Create Gateway Payment Failed.', errorType, details);
+  }
+};
+
+const processGatewayPayment = async ({payment}) => {
+  const order = await getOrderFromPayment({payment});
+  const verifiedPurchase = verifyOrder({order, payment});
+  return verifiedPurchase;
+};
+
+/**
+ * This is only used in the tests to simulate
+ * credit card failures.
+ *
+ * @param {object} options - Options to use.
+ * @param {object} options.order - A PayPal order object.
+ * @param {object} options.card - The payee's card.
```

Comment:
Not used.  Should doc body?

Reply:
yeah changed it to doc body here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/42a81daf80556f5ef82c3744e34cb7533e466d79